src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -263,28 +286,82 @@ overwrite_ops_t prepare_ops_list(
       assert(to_write.size() > 1);
       assert(back.addr + back.len ==
 	back.pin->get_key() + back.pin->get_length());
-      ops.to_remap.push_back(extent_to_remap_t::create_remap(
+      ops.to_remap.push_back(extent_to_remap_t::create_remap1(
 	std::move(back.pin),
 	back.addr - back.pin->get_key(),
 	back.len));
       ops.to_remove.pop_back();
     }
   }
 
-  // prepare to_insert
+  interval_set<uint64_t> pre_alloc_addr_removed, pre_alloc_addr_remapped;
+  if (delta_based_overwrite_max_extent_size) {
+    for (auto &r : ops.to_remove) {
+      if (r->get_val().is_absolute()) {
+	pre_alloc_addr_removed.insert(r->get_key(), r->get_length());
+
+      }
+    }
+    for (auto &r : ops.to_remap) {
+      if (r.pin && r.pin->get_val().is_absolute()) {
+	pre_alloc_addr_remapped.insert(r.pin->get_key(), r.pin->get_length());
+      }
+    }
+  }
+
+  // prepare to insert
+  extent_to_remap_list_t to_remap;
   for (auto &region : to_write) {
     if (region.is_data()) {
       visitted++;
       assert(region.to_write.has_value());
-      ops.to_insert.push_back(extent_to_insert_t::create_data(
-	region.addr, region.len, region.to_write));
+      int erased_num = 0;
+      if (pre_alloc_addr_removed.contains(region.addr, region.len) &&
+	  region.len <= delta_based_overwrite_max_extent_size) {
+	erased_num = std::erase_if(
+	  ops.to_remove,
+	  [&region, &to_remap](auto &r) {
+	    interval_set<uint64_t> range;
+	    range.insert(r->get_key(), r->get_length());
+	    if (range.contains(region.addr, region.len)) {
+	      to_remap.push_back(extent_to_remap_t::create_overwrite(
+		0, region.len, r->get_key(), r->get_length(), region.to_write));
```

Comment:
see comment https://github.com/ceph/ceph/pull/54036#issuecomment-1774375771

create_overwrite() should not be called if the target extent is cloned.

cc @xxhdx1985126 Looks the possible way is to either:
1. attach ref-counter to the returned LBAMapping.
2. track ref-counter in the stable logical cached extent at runtime.

Looks the 2nd way is more generic than 1, what do you think?

Same to L343 below.

Reply:
If I understand your suggestion correctly, "track ref-counter in the stable logical cached extent" means that a new refcount needs to be added to the logical cached extent, right?

If correct, how can we maintain the ref-counter after mount time? Should we do a full search to find all references for the extent?

Reply:
Can you take a look at the last commit I added? By using get_conf() to LBANode, we probably are able to get a reference counter for the snapped region. Am I missing something?

Reply:
> If I understand your suggestion correctly, "track ref-counter in the stable logical cached extent" means that a new refcount needs to be added to the logical cached extent, right?

Yes, I suggested to add the refcount to the logical cached extent.

But after another thought, that way may be challenging because different transactions may need to see different refcounts locally, but an extent can be shared among transactions. So we cannot directly store a transaction-local refcount in the shared extent, and the option 1 above can be simpler.

> Can you take a look at the last commit I added? By using get_conf() to LBANode, we probably are able to get a reference counter for the snapped region. Am I missing something?

Looks another lba lookup to get the laddr's refcount by `TM::get_ref()` is unnecessary.

The original region is retrieved from `TM::get_pin(s)`, where the refcount can be available if attatched to the returned `pin`, right?

Reply:
> The original region is retrieved from TM::get_pin(s), where the refcount can be available if attatched to the returned pin, right?

Right. I addressed your comment.

Reply:
Please refer to my reviews below.